Impl Read<u8> / Write<u8> for Serial

dbrgn opened the pull request:
Fixes #167.

I managed to implement this without having split tx/rx pins in the `Serial` struct thanks to the help of @Disasm. Instead, it uses a private trait (`UsartReadWrite`) to avoid having duplicated peripheral access logic. There should be no breaking change in this PR.

TheZoq2 commented:
Seems like a nice and simple change, good job! The last matrix message seems to suggest that you may want to make additional improvements. Is that the case or is it ready for merging?

dbrgn commented:
> The last matrix message seems to suggest that you may want to make additional improvements. Is that the case or is it ready for merging?

I couldn't figure out how to implement the change suggested by @Disasm, but maybe I misunderstood something...

To me, it seems better to keep the trait methods as associated functions without a &self reference, since they actually *are* static and don't require a reference to an instance.

Disasm commented:
Passing `&self` ensures ownership, but it is something we do not have in all of the cases anyway. I'm trying to improve this code locally it's dfficult to implement this way. So LGTM. Maybe I will improve this in the future.

Disasm commented:
Well, I have an update. Should I open a PR to the dbrgn's PR or should I PR my changes here after this PR gets merged?

dbrgn commented:
> Well, I have an update. Should I open a PR to the dbrgn's PR or should I PR my changes here after this PR gets merged?

If you want, you can also send me the patch (`git show > patch.txt`) and I'll add it to this PR :slightly_smiling_face:

Disasm commented:
Here we are: https://github.com/dbrgn/stm32f1xx-hal/pull/1

dbrgn commented:
Thanks! I merged your changes, will add a comment.

dbrgn commented on src/serial.rs:
```diff
@@ -357,112 +420,42 @@ macro_rules! hal {
 
                 fn read(&mut self) -> nb::Result<u8, Error> {
-                    $USARTX::read()
+                    unsafe { &*$USARTX::ptr() }.read()
```
Out of curiosity, what's the advantage of taking `&self` instead of using associated methods, if the new approach even requires an unsafe raw pointer dereference?

dbrgn replied:
Ah, the binary gets smaller (as mentioned in a comment [below](https://github.com/stm32-rs/stm32f1xx-hal/pull/171#issuecomment-575720942)). I guess that's a good tradeoff?

Disasm replied:
Originally, this unsafe raw pointer dereference was everywhere. Now we have it only when necessary.

dbrgn replied:
Ahh, you're right, I missed that! Much better indeed :slightly_smiling_face:

Disasm commented:
With this implementation only one copy of `UsartReadWrite::{read/write/flush}` is created in debug, as opposed to the original implementation with macros.

dbrgn commented:
I rebased the branch and squashed my fixup commits. I also integrated the "Strip confusing comment" commit into the commit "Deduplicate USART driver code".

From my side, this PR is now ready.

Disasm commented:
@TheZoq2 could you review this one more time?

TheZoq2 commented:
Looks great, thank you both for your work!

TheZoq2 commented:
Oh yea, just one minor thing. Would you mind adding a changelog entry before I merge?

dbrgn commented:
> Oh yea, just one minor thing. Would you mind adding a changelog entry before I merge?

Sure! Forgot about that. Done!

TheZoq2 commented:
Lovely!
